packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -0,0 +1,356 @@
+import UltraFeedRepo from "../../lib/ultraFeed/UltraFeedRepo";
+import {
+  FeedItemSourceType, UltraFeedResolverType, FeedItemRenderType, FeedItem,
+  FeedSpotlight, FeedFullPost, FeedCommentMetaInfo,
+  PreDisplayFeedComment,
+  FeedCommentsThread,
+  FeedCommentsThreadResolverType,
+  FeedPostResolverType
+} from "@/components/ultraFeed/ultraFeedTypes";
+import { filterNonnull } from "@/lib/utils/typeGuardUtils";
+import gql from 'graphql-tag';
+import { UltraFeedEvents } from '../collections/ultraFeedEvents/collection';
+import { bulkRawInsert } from '../manualMigrations/migrationUtils';
+
+export const ultraFeedGraphQLTypeDefs = gql`
+  type FeedPost {
+    _id: String!
+    postMetaInfo: JSON
+    post: Post
+  }
+
+  type FeedCommentThread {
+    _id: String!
+    commentMetaInfos: JSON
+    comments: [Comment]
+    post: Post                         
+  }
+
+  type FeedSpotlightItem {
+    _id: String!
+    spotlight: Spotlight
+  }
+
+  type UltraFeedQueryResults {
+    cutoff: Date
+    endOffset: Int!
+    results: [UltraFeedEntryType!]
+    sessionId: String
+  }
+
+  type UltraFeedEntryType {
+    type: String!
+    feedCommentThread: FeedCommentThread
+    feedPost: FeedPost
+    feedSpotlight: FeedSpotlightItem
+  }
+
+  extend type Query {
+    UltraFeed(
+      limit: Int,
+      cutoff: Date,
+      offset: Int,
+      sessionId: String
+    ): UltraFeedQueryResults!
+  }
+`
+const SOURCE_WEIGHTS = {
+  postThreads: 20,
+  commentThreads: 40,
+  spotlights: 1,
+};
+
+type UsedFeedItemSourceType = Extract<FeedItemSourceType, "postThreads" | "commentThreads" | "spotlights">;
+interface WeightedSource {
+  weight: number;
+  items: FeedItem[];
+  renderAsType: FeedItemRenderType;
+}
+
+type SampledItem = { renderAsType: "feedCommentThread", feedCommentThread: FeedCommentsThread }
+                 | { renderAsType: "feedPost", feedPost: FeedFullPost }
+                 | { renderAsType: "feedSpotlight", feedSpotlight: FeedSpotlight };
+
+function weightedSample(
```

Comment:
I haven't looked too carefully at what's going on here but my experience on April Fools was that it's much easier to mess up weighted sampling than you'd expect

Reply:
noted. I think I can work on this more once anything vaguely working is shipped to admins